aptos-move/aptos-aggregator/src/delayed_change.rs

```diff
@@ -120,6 +122,148 @@ impl<I: Copy + Clone> DelayedChange<I> {
+        use DelayedApplyEntry::*;
+
+        match self {
+            AggregatorDelta { .. } => None,
+            SnapshotDelta {
+                base_aggregator, ..
+            } => Some(ApplyBase::Previous(*base_aggregator)),
+            SnapshotDerived { base_snapshot, .. } => Some(ApplyBase::Current(*base_snapshot)),
+        }
+    }
+
+    pub fn get_apply_base_id(&self, self_id: &I) -> ApplyBase<I> {
+        self.get_apply_base_id_option()
+            .unwrap_or(ApplyBase::Previous(*self_id))
+    }
+
+    pub fn apply_to_base(
+        &self,
+        base_value: DelayedFieldValue,
+    ) -> Result<DelayedFieldValue, PanicOr<DelayedFieldsSpeculativeError>> {
+        use DelayedApplyEntry::*;
+
+        Ok(match self {
+            AggregatorDelta { delta } => {
+                DelayedFieldValue::Aggregator(delta.apply_to(base_value.into_aggregator_value()?)?)
+            },
+            SnapshotDelta { delta, .. } => {
+                DelayedFieldValue::Snapshot(delta.apply_to(base_value.into_aggregator_value()?)?)
+            },
+            SnapshotDerived { formula, .. } => {
+                DelayedFieldValue::Derived(formula.apply_to(base_value.into_snapshot_value()?))
+            },
+        })
+    }
+}
+
+impl<I: Copy + Clone> DelayedEntry<I> {
+    // When squashing a new change on top of the old one, sometimes we need to know the change
+    // from a different AggregatorID to be able to merge them together.
+    // In particular SnapshotDelta represents a change from the aggregator at the beginning of the transaction,
+    // and squashing changes where the aggregator will be at the beginning of the transaction.
+    // For example, let’s say we have two change sets that we need to squash:
+    // change1: agg1 -> Delta(+3)
+    // change2: agg1 -> Delta(+6), snap1 -> (base=agg1, Delta(+2))
+    // the correct squashing of snapshot depends on the change for the base aggregator. I.e. the correct output would be:
+    // agg1 -> Delta(+9), snap(base=agg1, Delta(+5))
+    pub fn get_merge_dependent_id(&self) -> Option<I> {
+        use DelayedApplyEntry::*;
+        use DelayedEntry::*;
+
+        match self {
+            // Only SnapshotDelta merging logic depends on current aggregator change
+            Apply(SnapshotDelta {
+                base_aggregator, ..
+            }) => Some(*base_aggregator),
+            Create(_) | Apply(AggregatorDelta { .. } | SnapshotDerived { .. }) => None,
+        }
+    }
+
+    /// Applies next AggregatorChange on top of the previous state.
+    /// If get_merge_dependent_id() returns some add, prev_change passed in should be
+    /// for that particular AggregatorID.
+    /// If get_merge_dependent_id() returns None, prev_change is required to be for the
+    /// same AggregatorID as next_change.
+    pub fn merge_two_changes(
+        prev_change: Option<&DelayedEntry<I>>,
+        next_change: &DelayedEntry<I>,
```

Comment:
So, there are two `merge_two_changes` functions -- one for DelayedEntry and another for DelayedChange?

Reply:
cleaning up, DelayedEntry's one can be removed